File: HW/HW10/HW10.java
import java.util.ArrayList;
import java.util.LinkedList;
import java.util.List;
import java.util.Queue;

public class HW10 {
	public static List<List<Integer>> levelOrder(TreeNode root) {
        List<List<Integer>> result = new ArrayList<>();
        if (root == null) {
            return result;
        }

        Queue<TreeNode> queue = new LinkedList<>();
        queue.offer(root);

        while (!queue.isEmpty()) {
            int levelSize = queue.size();
            List<Integer> levelNodes = new ArrayList<>();

            for (int i = 0; i < levelSize; i++) {
                TreeNode node = queue.poll();
                levelNodes.add(node.val);

                if (node.left != null) {
                    queue.offer(node.left);
                }
                if (node.right != null) {
                    queue.offer(node.right);
                }
            }

            result.add(levelNodes);
        }

        return result;
    }
    
	 public static void main(String[] args) {
		 TreeNode node1 = new TreeNode(1);
		 TreeNode node5 = new TreeNode(5);
		 TreeNode node9 = new TreeNode(9);
		 TreeNode node3 = new TreeNode(3, node1, null);
		 TreeNode node8 = new TreeNode(8, node5, node9);
		 TreeNode root = new TreeNode(4, node3, node8);
		 
		 System.out.println(levelOrder(root));
	 }
}

File: HW/HW10/HW10.py
from collections import deque
from typing import Optional
from typing import List


class TreeNode(object):

  def __init__(self, val=0, left=None, right=None):

    self.val = val

    self.left = left

    self.right = right


def levelOrder(root: Optional[TreeNode]) -> List[List[int]]:
  result = []
  if root is None:
    return result

  queue = deque()
  queue.append(root)

  while queue:
    level_size = len(queue)
    level_nodes = []

    for _ in range(level_size):
      node = queue.popleft()
      level_nodes.append(node.val)

      if node.left:
        queue.append(node.left)
      if node.right:
        queue.append(node.right)

    result.append(level_nodes)

  return result


node1 = TreeNode(1)
node5 = TreeNode(5)
node9 = TreeNode(9)
node3 = TreeNode(3, node1)
node8 = TreeNode(8, node5, node9)
root = TreeNode(4, node3, node8)

print(levelOrder(root))
